From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Tue, 13 Oct 2026 00:50:47 -0700
Subject: [PATCH] jscript: Expected ';', prefix ++/-- after a line break, Error
 objects only

Two statements without a separator on one line (also a byte order mark
splitting a word) fail with "Expected ';'" instead of "Syntax error". ++ and --
after a line break are only prefix operators, so "a\n-->0" is a syntax error
and "a\n++b" is two statements. Only Error instances give their number and
description to the host's error info (and the user locale in wReserved); any
other thrown object is "Exception thrown and not caught". The todo marks of
run.c that pass now are removed.

diff --git a/dlls/jscript/engine.c b/dlls/jscript/engine.c
index c95fcac..bb297bf 100644
--- a/dlls/jscript/engine.c
+++ b/dlls/jscript/engine.c
@@ -1211,13 +1211,17 @@ static void set_error_value(script_ctx_t *ctx, jsval_t value)
     ei->valid_value = TRUE;
     ei->value = value;
 
+    #ifdef SG_MUTANT_JS_ANYOBJ_ERROR
     if(is_object_instance(value) && (obj = to_jsdisp(get_object(value)))) {
+#else
+    if(is_object_instance(value) && (obj = to_jsdisp(get_object(value))) && obj->builtin_info->class == JSCLASS_ERROR) {
+#endif
         UINT32 number;
         jsstr_t *str;
         jsval_t v;
         HRESULT hres;
 
-        /* FIXME: We should check if object is an error instance */
+        ei->error_object = TRUE;
 
         hres = jsdisp_propget_name(obj, L"number", &v);
         if(SUCCEEDED(hres)) {
diff --git a/dlls/jscript/engine.h b/dlls/jscript/engine.h
index c1a4293..4649b79 100644
--- a/dlls/jscript/engine.h
+++ b/dlls/jscript/engine.h
@@ -252,6 +252,7 @@ struct _jsexcept_t {
     HRESULT error;
 
     BOOL valid_value;
+    BOOL error_object;  /* the thrown value is an Error instance */
     jsval_t value;
 
     jsstr_t *source;
diff --git a/dlls/jscript/jscript.c b/dlls/jscript/jscript.c
index c2429c1..404c081 100644
--- a/dlls/jscript/jscript.c
+++ b/dlls/jscript/jscript.c
@@ -253,6 +253,9 @@ static HRESULT WINAPI JScriptError_GetExceptionInfo(IActiveScriptError *iface, E
 
     memset(excepinfo, 0, sizeof(*excepinfo));
     excepinfo->scode = This->ei.error;
+    /* an Error object thrown by the script carries the user's locale */
+    if(This->ei.error_object)
+        excepinfo->wReserved = LOWORD(GetUserDefaultLCID());
     if(This->ei.source)
         jsstr_to_bstr(This->ei.source, &excepinfo->bstrSource);
     if(This->ei.message)
@@ -315,6 +318,7 @@ static const IActiveScriptErrorVtbl JScriptErrorVtbl = {
 void reset_ei(jsexcept_t *ei)
 {
     ei->error = S_OK;
+    ei->error_object = FALSE;
     if(ei->valid_value) {
         jsval_release(ei->value);
         ei->valid_value = FALSE;
diff --git a/dlls/jscript/lex.c b/dlls/jscript/lex.c
index ee29bff..33fccbf 100644
--- a/dlls/jscript/lex.c
+++ b/dlls/jscript/lex.c
@@ -670,7 +670,8 @@ static int next_token(parser_ctx_t *ctx, unsigned *loc, void *lval)
             switch(*ctx->ptr) {
             case '+':  /* ++ */
                 ctx->ptr++;
-                return tINC;
+                /* ++ and -- after a line terminator can only be prefix operators */
+                return ctx->nl ? tINCNL : tINC;
             case '=':  /* += */
                 ctx->ptr++;
                 *(int*)lval = EXPR_ASSIGNADD;
@@ -689,7 +690,7 @@ static int next_token(parser_ctx_t *ctx, unsigned *loc, void *lval)
                     ctx->ptr++;
                     return tHTMLCOMMENT;
                 }
-                return tDEC;
+                return ctx->nl ? tDECNL : tDEC;
             case '=':  /* -= */
                 ctx->ptr++;
                 *(int*)lval = EXPR_ASSIGNSUB;
diff --git a/dlls/jscript/parser.y b/dlls/jscript/parser.y
index 949d913..d6f6233 100644
--- a/dlls/jscript/parser.y
+++ b/dlls/jscript/parser.y
@@ -30,6 +30,11 @@ static int parser_error(unsigned*,parser_ctx_t*,const char*);
 static void set_error(parser_ctx_t*,unsigned,HRESULT);
 static BOOL explicit_error(parser_ctx_t*,void*,WCHAR);
 static BOOL allow_auto_semicolon(parser_ctx_t*);
+#ifdef SG_MUTANT_JS_SYNTAX_ONLY
+#define JS_E_MISSING_SEMICOLON_MUT JS_E_SYNTAX
+#else
+#define JS_E_MISSING_SEMICOLON_MUT JS_E_MISSING_SEMICOLON
+#endif
 
 static literal_t *new_string_literal(parser_ctx_t*,jsstr_t*);
 static literal_t *new_null_literal(parser_ctx_t*);
@@ -163,7 +168,7 @@ static expression_t *new_prop_and_value_expression(parser_ctx_t*,property_list_t
 %token <identifier> kBREAK kCASE kCATCH kCONST kCONTINUE kDEFAULT kDELETE kDO kELSE kFUNCTION kIF kFINALLY kFOR
 %token <identifier> kGET kIN kLET kSET kINSTANCEOF kNEW kNULL kRETURN kSWITCH kTHIS kTHROW kTRUE kFALSE
 %token <identifier> kTRY kTYPEOF kVAR kVOID kWHILE kWITH
-%token tANDAND tOROR tINC tDEC tHTMLCOMMENT kDIVEQ kDCOL
+%token tANDAND tOROR tINC tDEC tINCNL tDECNL tHTMLCOMMENT kDIVEQ kDCOL
 
 /* tokens */
 %token <identifier> tIdentifier
@@ -735,6 +740,10 @@ UnaryExpression
                                 { $$ = new_unary_expression(ctx, EXPR_TYPEOF, $2); }
         | tINC UnaryExpression  { $$ = new_unary_expression(ctx, EXPR_PREINC, $2); }
         | tDEC UnaryExpression  { $$ = new_unary_expression(ctx, EXPR_PREDEC, $2); }
+        | tINCNL UnaryExpression
+                                { $$ = new_unary_expression(ctx, EXPR_PREINC, $2); }
+        | tDECNL UnaryExpression
+                                { $$ = new_unary_expression(ctx, EXPR_PREDEC, $2); }
         | '+' UnaryExpression   { $$ = new_unary_expression(ctx, EXPR_PLUS, $2); }
         | '-' UnaryExpression   { $$ = new_unary_expression(ctx, EXPR_MINUS, $2); }
         | '~' UnaryExpression   { $$ = new_unary_expression(ctx, EXPR_BITNEG, $2); }
@@ -950,7 +959,7 @@ BooleanLiteral
 
 semicolon_opt
         : ';'
-        | error                 { if(!allow_auto_semicolon(ctx)) {YYABORT;} else { ctx->hres = S_OK; ctx->error_loc = -1; } }
+        | error                 { if(!allow_auto_semicolon(ctx)) {set_error(ctx, @$, JS_E_MISSING_SEMICOLON_MUT); YYABORT;} else { ctx->hres = S_OK; ctx->error_loc = -1; } }
 
 left_bracket
         : '('
diff --git a/dlls/jscript/tests/run.c b/dlls/jscript/tests/run.c
index 106fd8c..ee8246f 100644
--- a/dlls/jscript/tests/run.c
+++ b/dlls/jscript/tests/run.c
@@ -187,7 +187,7 @@ DEFINE_EXPECT(BindHandler);
 #define DISPID_TESTOBJ_WITHPROP     0x2002
 #define DISPID_TESTOBJ_TOLOCALESTR  0x2003
 
-#define JS_E_OUT_OF_MEMORY 0x800a03ec
+#define JS_E_MISSING_SEMICOLON 0x800a03ec
 #define JS_E_INVALID_CHAR 0x800a03f6
 
 static BOOL strict_dispid_check, testing_expr;
@@ -2310,6 +2310,7 @@ static HRESULT parse_htmlscript(const WCHAR *script_str)
 #define ERROR_TODO_SCODE        0x0002
 #define ERROR_TODO_DESCRIPTION  0x0004
 #define ERROR_TODO_HELPFILE     0x0008
+#define ERROR_TODO_LCID         0x0010
 
 static void test_error_reports(void)
 {
@@ -2345,8 +2346,7 @@ static void test_error_reports(void)
             L"Microsoft JScript compilation error",
             L"Expected ')'",
             NULL, 0,
-            L"-->0) a=5;",
-            ERROR_TODO_PARSE
+            L"-->0) a=5;"
         },
         {
             L"new 3;",
@@ -2439,8 +2439,8 @@ static void test_error_reports(void)
             NULL,
             NULL, 0,
             NULL,
-            FALSE,
-            0x409
+            ERROR_TODO_LCID,
+            TRUE
         },
         {
             L" throwEI(-2147467259 /* E_FAIL */);",
@@ -2540,8 +2540,7 @@ static void test_error_reports(void)
             L"Microsoft JScript runtime error",
             L"Exception thrown and not caught",
             NULL, 0,
-            NULL,
-            ERROR_TODO_SCODE | ERROR_TODO_DESCRIPTION
+            NULL
         },
         {
             L"f(1\n,\n2,\n ,,3\n);\n",
@@ -2653,7 +2652,7 @@ static void test_error_reports(void)
             todo_wine_if(tests[i].todo_flags & ERROR_TODO_SCODE)
             ok(ei.scode == tests[i].error, "[%u] scode = %08lx, expected %08lx\n", i, ei.scode, tests[i].error);
             ok(ei.wCode == 0, "wCode = %x\n", ei.wCode);
-            todo_wine_if(tests[i].reserved_lcid)
+            todo_wine_if(tests[i].todo_flags & ERROR_TODO_LCID)
             ok(ei.wReserved == (tests[i].reserved_lcid ? GetUserDefaultLCID() : 0), "[%u] wReserved = %x expected %lx\n",
                i, ei.wReserved, (tests[i].reserved_lcid ? GetUserDefaultLCID() : 0));
             if (is_lang_english())
@@ -3964,10 +3963,10 @@ static void run_bom_tests(void)
     struct bom_test bom_tests[] = {
         {L"var a = 1; reportSuccess();", S_OK},
         {L"\xfeffvar a = 1; reportSuccess();", S_OK},
-        {L"v\xfeff" "ar a = 1; reportSuccess();", JS_E_OUT_OF_MEMORY},
+        {L"v\xfeff" "ar a = 1; reportSuccess();", JS_E_MISSING_SEMICOLON},
         {L"var\xfeff a = 1; reportSuccess();", S_OK},
         {L"var a = 1; \xfeffreportSuccess();", S_OK},
-        {L"var a = 1; report\xfeffSuccess();", JS_E_OUT_OF_MEMORY},
+        {L"var a = 1; report\xfeffSuccess();", JS_E_MISSING_SEMICOLON},
         {L"var a = 1; reportSuccess\xfeff();", S_OK},
         {L"var a = 1; reportSuccess(\xfeff);", S_OK},
         {L"var a =\xfeff 1; reportSuccess(\xfeff);", S_OK},
@@ -3994,7 +3993,7 @@ static void run_bom_tests(void)
         {
              src = SysAllocString(bom_tests[i].str);
              hres = parse_script(SCRIPTITEM_GLOBALMEMBERS, src);
-             todo_wine ok(hres == bom_tests[i].hres, "test %s returned with %08lx\n", wine_dbgstr_w(src), hres);
+             ok(hres == bom_tests[i].hres, "test %s returned with %08lx\n", wine_dbgstr_w(src), hres);
              SysFreeString(src);
         }
     }
